subsys/bluetooth/mesh/proxy_cli.c

```diff
@@ -0,0 +1,314 @@
+#include "gatt_cli.h"
+#include "proxy_msg.h"
+
+static struct bt_mesh_proxy_server {
+	struct bt_mesh_proxy_role *srv;
+	bool link_opened;
+} servers[CONFIG_BT_MAX_CONN];
+
+static uint16_t allow_net_idx[CONFIG_BT_MAX_CONN] = {
+	[0 ... (CONFIG_BT_MAX_CONN - 1)] = BT_MESH_KEY_UNUSED,
+};
+
+static uint16_t *find_net_idx(uint16_t net_idx)
+{
+	for (int i = 0; i < ARRAY_SIZE(allow_net_idx); i++) {
+		if (allow_net_idx[i] == net_idx) {
+			return &allow_net_idx[i];
+		}
+	}
+
+	return NULL;
+}
+
+static void buf_send_end(struct bt_conn *conn, void *user_data)
+{
+	struct net_buf *buf = user_data;
+
+	net_buf_unref(buf);
+}
+
+bool bt_mesh_proxy_client_relay(struct net_buf *buf)
+{
+	bool relayed = false;
+	int i, err;
+
+	for (i = 0; i < ARRAY_SIZE(servers); i++) {
+		struct bt_mesh_proxy_server *server = &servers[i];
+
+		NET_BUF_SIMPLE_DEFINE(msg, 32);
+
+		if (!server->srv || !server->link_opened) {
+			continue;
+		}
+
+		/* Proxy PDU sending modifies the original buffer,
+		 * so we need to make a copy.
+		 */
+		net_buf_simple_reserve(&msg, 1);
+		net_buf_simple_add_mem(&msg, buf->data, buf->len);
+
+		err = bt_mesh_proxy_msg_send(server->srv->conn, BT_MESH_PROXY_NET_PDU,
+					     &msg, buf_send_end, net_buf_ref(buf));
+
+		bt_mesh_adv_send_start(0, err, BT_MESH_ADV(buf));
+		if (err) {
+			BT_ERR("Failed to send proxy message (err %d)", err);
+
+			/* If segment_and_send() fails the buf_send_end() callback will
+			 * not be called, so we need to clear the user data (net_buf,
+			 * which is just opaque data to segment_and send) reference given
+			 * to segment_and_send() here.
+			 */
+			net_buf_unref(buf);
+			continue;
+		}
+
+		relayed = true;
+	}
+
+	return relayed;
+}
+
+static int subnet_find_cb(struct bt_mesh_subnet *sub, void *cb_data)
+{
+	struct net_buf_simple *buf = (struct net_buf_simple *)cb_data;
+
+	if (!memcmp(sub->keys[0].net_id, buf->data, 8)) {
+		return 1;
+	}
+
+	if (!bt_mesh_subnet_has_new_key(sub)) {
+		return 0;
+	}
+
+	if (!memcmp(sub->keys[1].net_id, buf->data, 8)) {
+		return 1;
+	}
+
+	return 0;
+}
+
+static void proxy_msg_recv(struct bt_mesh_proxy_role *role)
+{
+	switch (role->msg_type) {
+	case BT_MESH_PROXY_NET_PDU:
+		BT_DBG("Mesh Network PDU");
+		bt_mesh_net_recv(&role->buf, 0, BT_MESH_NET_IF_PROXY);
+		break;
+	case BT_MESH_PROXY_BEACON:
+		BT_DBG("Mesh Beacon PDU");
+		bt_mesh_beacon_recv(&role->buf);
+		break;
+	case BT_MESH_PROXY_CONFIG:
+		BT_DBG("Mesh Configuration PDU");
+		/* TODO */
+		break;
+	default:
+		BT_WARN("Unhandled Message Type 0x%02x", role->msg_type);
+		break;
+	}
+}
+
+static void proxy_connected(struct bt_conn *conn)
+{
+	servers[bt_conn_index(conn)].srv =
+		bt_mesh_proxy_role_setup(conn, bt_mesh_gatt_send,
+					 proxy_msg_recv);
+
+}
+
+static void proxy_link_open(struct bt_conn *conn)
+{
+	servers[bt_conn_index(conn)].link_opened = true;
+}
+
+static void proxy_disconnected(struct bt_conn *conn)
+{
+	bt_mesh_proxy_role_cleanup(servers[bt_conn_index(conn)].srv);
```

Comment:
gatt_cli.c should handle this itself.

Reply:
I don't think so, because setup is set here, and cleanup should also be handled here? Isn't it?